source/fab/steps/cleanup_prebuilds.py

```diff
@@ -14,112 +14,113 @@
-            oldest_ts_allowed = most_recent_ts - self.older_than
-
-            for f, ts in prebuilds_ts.items():
-                if ts < oldest_ts_allowed:
-                    # don't delete if it's still current
-                    if f in current_files:
-                        logger.debug(f"old file is still current {f}")
-                        continue
-                    logger.info(f"old file {f}")
-                    to_delete.add(f)
-
-        return to_delete
-
-    def by_version_age(self, prebuilds_ts: Dict[Path, datetime], current_files: Iterable[Path]) -> Set[Path]:
-        to_delete = set()
-
-        if self.n_versions:
-            # group prebuild files by originating artefact, <stem>.*.<suffix>
-            pb_file_groups = get_prebuild_file_groups(prebuilds_ts.keys())
-
-            # delete the n oldest in each group
-            for pb_group in pb_file_groups.values():
-                by_age = sorted(pb_group, key=lambda f: prebuilds_ts[f], reverse=True)
-                for f in by_age[self.n_versions:]:
-                    # don't delete if it's still current
-                    if f in current_files:
-                        logger.debug(f"old version is still current {f}")
-                        continue
-                    logger.debug(f"old version {f}")
-                    to_delete.add(f)
-
-        return to_delete
+    # If the user has not specified any cleanup parameters, we default to a hard cleanup.
+    if not n_versions and not older_than:
+        if all_unused not in [None, True]:
+            raise ValueError(f"unexpected value for all_unused: '{all_unused}'")
+        all_unused = True
+
+    # if we're doing a hard cleanup, there's no point providing the softer options
+    if all_unused and (n_versions or older_than):
+        raise ValueError("n_versions or older_than should not be specified with all_unused")
+
+    num_removed = 0
+
+    # see what's in the prebuild folder
+    prebuild_files = list(file_walk(config.prebuild_folder))
+    if not prebuild_files:
+        logger.info('no prebuild files found')
+
+    elif all_unused:
+        num_removed = remove_all_unused(
+            found_files=prebuild_files, current_files=config._artefact_store[CURRENT_PREBUILDS])
+
+    else:
+        # get the file access time for every artefact
+        prebuilds_ts = \
+            dict(zip(prebuild_files, run_mp(config, prebuild_files, get_access_time)))  # type: ignore
+
+        # work out what to delete
+        to_delete = by_age(older_than, prebuilds_ts, current_files=config._artefact_store[CURRENT_PREBUILDS])
+        to_delete |= by_version_age(n_versions, prebuilds_ts, current_files=config._artefact_store[CURRENT_PREBUILDS])
+
+        # delete them all
+        run_mp(config, to_delete, os.remove)
+        num_removed = len(to_delete)
+
+    logger.info(f'removed {num_removed} prebuild files')
+    config._artefact_store[CLEANUP_COUNT] = num_removed
```

Comment:
We've added this because we can no longer detect the cleanup step from a list of steps.